local/compose.go

```diff
@@ -65,6 +70,103 @@ func (s *composeService) Build(ctx context.Context, project *types.Project) erro
 	return s.build(ctx, project, opts)
 }
 
+func (s *composeService) Push(ctx context.Context, project *types.Project) error {
+	configFile, err := config.Load(config.Dir())
+	if err != nil {
+		return err
+	}
+	eg, ctx := errgroup.WithContext(ctx)
+
+	info, err := s.apiClient.Info(ctx)
+	if err != nil {
+		return err
+	}
+	if info.IndexServerAddress == "" {
+		info.IndexServerAddress = registry.IndexServer
+	}
+
+	for _, service := range project.Services {
+		if service.Build == nil {
+			continue
+		}
+		service := service
+		eg.Go(func() error {
+			w := progress.ContextWriter(ctx)
+
+			ref, err := reference.ParseNormalizedNamed(service.Image)
+			if err != nil {
+				return err
+			}
+
+			repoInfo, err := registry.ParseRepositoryInfo(ref)
+			if err != nil {
+				return err
+			}
+
+			key := repoInfo.Index.Name
+			if repoInfo.Index.Official {
+				key = info.IndexServerAddress
+			}
+			authConfig, err := configFile.GetAuthConfig(key)
+			if err != nil {
+				return err
+			}
+
+			buf, err := json.Marshal(authConfig)
+			if err != nil {
+				return err
+			}
+
+			stream, err := s.apiClient.ImagePush(ctx, service.Image, moby.ImagePushOptions{
+				RegistryAuth: base64.URLEncoding.EncodeToString(buf),
+			})
+			if err != nil {
+				return err
+			}
+			dec := json.NewDecoder(stream)
+			for {
+				var jm jsonmessage.JSONMessage
+				if err := dec.Decode(&jm); err != nil {
+					if err == io.EOF {
+						break
+					}
+					return err
+				}
+				toProgressEvent(service.Name, jm, w)
+			}
+			return nil
+		})
+	}
+	return eg.Wait()
+}
+
+func toProgressEvent(prefix string, jm jsonmessage.JSONMessage, w progress.Writer) {
+	if jm.ID == "" {
+		return
+	}
+	var (
+		text   string
+		status = progress.Working
+	)
+
+	if jm.Status == "Pull complete" || jm.Status == "Already exists" {
+		status = progress.Done
+	}
+	if jm.Error != nil {
+		status = progress.Error
+		text = jm.Error.Message
+	}
+	if jm.Progress != nil {
+		text = jm.Progress.String()
+	}
+	w.Event(progress.Event{
+		ID:         fmt.Sprintf("%s: %s", prefix, jm.ID),
```

Comment:
Nit: might be nicer adding "Pushing" as `fmt.Sprintf("Pushing %s: %s", prefix, jm.ID)` ,
```
[+] Running 0/24
 ⠴ web: d4852371a647 Layer already exists
 ⠴ web: 8c0845160e01 Layer already exists
 ⠴ words: 80f0fd999f52 Mounted from gtardif/sentences-api-testpush    
```

Reply:
Right, moby json.progress events hardly match the progresswriter model :P
By the way, I also noticed we don't stop when such an error happens (docker-compose has an option to ignore pull failures, we could do the same)